kernel/src/sched/sched_class/fair.rs

```diff
@@ -0,0 +1,257 @@
+
+use alloc::{collections::btree_set::BTreeSet, sync::Arc};
+use core::{
+    cmp,
+    ops::Bound,
+    sync::atomic::{AtomicU64, Ordering::*},
+};
+
+use ostd::{
+    cpu::{num_cpus, CpuId},
+    task::scheduler::UpdateFlags,
+};
+
+use super::{
+    time::{base_slice_clocks, min_period_clocks},
+    SchedClassRq,
+};
+use crate::{sched::priority::Nice, thread::Thread};
+
+const WEIGHT_0: u64 = 1024;
+pub const fn nice_to_weight(nice: Nice) -> u64 {
+    // Calculated by the formula below:
+    //
+    //     weight = 1024 * 1.25^(-nice)
+    //
+    // We propose that every increment of the nice value results
+    // in 12.5% change of the CPU load weight.
+    const FACTOR_NUMERATOR: u64 = 5;
+    const FACTOR_DENOMINATOR: u64 = 4;
+
+    match nice.range().get() {
+        0 => WEIGHT_0,
+        nice @ 1.. => {
+            let numerator = FACTOR_DENOMINATOR.pow(nice as u32);
+            let denominator = FACTOR_NUMERATOR.pow(nice as u32);
+            WEIGHT_0 * numerator / denominator
+        }
+        nice => {
+            let numerator = FACTOR_NUMERATOR.pow((-nice) as u32);
+            let denominator = FACTOR_DENOMINATOR.pow((-nice) as u32);
+            WEIGHT_0 * numerator / denominator
+        }
+    }
+}
+
+/// The scheduling entity for the FAIR scheduling class.
+///
+/// The structure contains 2 significant indications:
+/// `vruntime` & `period_start`.
+///
+/// # `vruntime`
+///
+/// The vruntime (virtual runtime) is calculated by the formula:
+///
+///     vruntime += (cur - start) * WEIGHT_0 / weight
+///
+/// and a thread with a lower vruntime gains a greater privilege to be
+/// scheduled, making the whole run queue balanced on vruntime (thus FAIR).
+///
+/// # Scheduling periods
+///
+/// Scheduling periods is designed to calculate the time slice for each threads.
+///
+/// The time slice for each threads is calculated by the formula:
+///
+///     time_slice = period * weight / total_weight
+///
+/// where `total_weight` is the sum of all weights in the run queue including
+/// the current thread and [`period`](FairClassRq::period) is calculated
+/// regarding the number of running threads.
+///
+/// When a thread's time slice is exhausted, it will be put back to the
+/// run queue.
+
+#[derive(Debug)]
+pub struct FairAttr {
+    weight: AtomicU64,
+    vruntime: AtomicU64,
+}
+
+impl FairAttr {
+    pub fn new(nice: Nice) -> Self {
+        FairAttr {
+            weight: nice_to_weight(nice).into(),
+            vruntime: Default::default(),
+        }
+    }
+
+    pub fn update(&self, nice: Nice) {
+        self.weight.store(nice_to_weight(nice), Relaxed);
+    }
+
+    fn tick(
+        &self,
+        now: u64,
+        start: u64,
+        period_start: u64,
+        total_weight: u64,
+        period: u64,
+    ) -> bool {
+        let weight = self.weight.load(Relaxed);
+
+        // Update the vruntime.
+        self.vruntime
+            .fetch_add((now - start) * WEIGHT_0 / weight, Relaxed);
```

Comment:
I am skeptical about the correctness of the current CFS implementation because of how `vruntime` and `period_start` values are updated and used.

First of all, after a task of `SCHED_OTHER` is spawned, the vruntime of this task is increased forever. Imagine this scenario: we have a task A that keeps running for a year. Then, a new task B is spawned. So the vruntime of task A will be much, much larger than that of task B. This means it would take a very loooong time before task A can be scheduled before task B. Even if task A is given the least nice value! Even if task A is I/O bound!

Second, the CFS algorithm is supposed to identify CPU-bound and I/O-bound tasks automatically by using vruntime and favor I/O-bound tasks over CPU-bound ones to reduce the latency of I/O. But the current CFS implementation does not behave this way. Consider the following situation: a CPU-bound task is the only running task on a CPU core. Then, an I/O-bound task wakes up and gets enqueued into a runqueue of a CFS scheduler. The I/O-bound task must have a small vruntime because it consumes lesser CPU time than the CPU-bound task. My understanding is that the expected behavior of CFS is to let the I/O-bound task (with a low vruntime) preempt the currently-running CPU-bound task (with a high vruntime) at the latest convenient moment (e.g., a scheduler tick).

```
    fn tick(
        &self,
        now: u64,
        start: u64,
        period_start: u64,
        total_weight: u64,
        period: u64,
    ) -> bool {
        // ...

        let slice = period * weight / total_weight;
        now - period_start > slice
}
```

But according to the current implementation (above), a preemption will only happen if the current running task consumes its assigned time slice. In other words, the I/O-bound task won't be able to preempt the CPU-bound one.

Third, I am confused with the semantics of the `period_start` attribute. Currently, this value of `period_start` is set when a task is _enqueued_ into a local runqueue.  (So this `period_start` should be called `enqueue_timestamp`.)

```rust
    fn enqueue_thread(&mut self, thread: &Arc<Thread>) {
        let attr = thread.sched_attr();

        attr.period_start.store(sched_clock(), Relaxed);
```

Then, the aforementioned tick method check if `now - period_start > slice`. Why do we want to calculate the elapsed time since the task is enqueued? This is NOT the consumed CPU time of this task. And then compare it with the assigned time slice? The code does not make sense to me at all.

Reply:
For possible starvation, yes, you're right. I've implemented a new mechanism `min_vruntime` to keep the values not too distant.

For responsiveness on I/O bound tasks, I've added another condition to trigger preemption:

    cur.vruntime >= rq.min_vruntime + normalized_slice

I've also refactored `tick` to ensure readability.

For `period_start`, yes, you're right. I've updated the period runtime counting algorithm.

Reply:
I have a working understanding of task scheduling, I am not an expert, especially when it comes to the intricacies of CFS. Implementing a robust and correct scheduling algorithm requires a level of detail that exceeds the time and resources I can currently dedicate. Therefore, I deeply value your expertise and professionalism in ensuring the accuracy and reliability of this implementation.

When I conduct code reviews, I do so under the assumption that the code is fundamentally correct, with the expectation that bugs are the exception rather than the norm. To this end, could I kindly ask you to perform an additional thorough review of the implementation? It has been three rounds of code review for this PR and I really hope we can merge it soon.

Reply:
I'm sorry for the trouble. I have been busy in the last two weeks, so the amount of time I've spent improving the quality of the PR is less than before. I hope it'll be merged soon from the bottom of my heart.